### ISSUE TITLE
Confusing behavior when getting degree of a node not in the graph

### ISSUE THREAD

**Issue by jfinkels** [networkx/networkx#2347]
If I ask for the degree of a node not in the graph (or directed graph), the method returns a generator object:
```sh
$ python -c "import networkx; print(networkx.Graph().degree(0))"
<generator object Graph.degree.<locals>.d_iter at 0x7fcfc64230a0>
```
I expected that this would have raised an exception, not return a generator:
```sh
$ python -c "import networkx; networkx.Graph().degree(0)"
networkx.exception.NodeNotFound: 0 is not in the graph
```

**Comment by dschult** [comment]
I think we should use views instead of generators for degree/edges/nodes. 
Then this issue might work better. #1943 is one such attempt.

**Comment by hagberg** [comment]
This produces plenty of errors though:

```
$ python -c "import networkx; print(list(networkx.Graph().degree(0)))"
```

**Comment by jfinkels** [comment]
@hagberg of course, but if I, the user, am asking a graph for the degree of a single node, why would I be calling `list()` on the return value?

**Comment by dschult** [comment]
Fixed by #2458 although the error message is about an invalid nbunch instead of a node because we expect an nbunch input to G.degree. Still, it raises an error now as it should.
